bezier-rs/lib/src/lib.rs

```diff
@@ -853,6 +838,159 @@ impl Bezier {
+		let ArcsOptions { maximize_arcs, error, max_iterations } = arcs_options;
+		match maximize_arcs {
+			MaximizeArcs::Automatic => {
+				let (auto_arcs, final_low_t) = self.arcs_helper(0., 1., error, max_iterations, true);
+				if final_low_t != 1. {
+					[
+						auto_arcs,
+						self.split(final_low_t)[1].arcs(ArcsOptions {
+							maximize_arcs: MaximizeArcs::Off,
+							error,
+							max_iterations,
+						}),
+					]
+					.concat()
+				} else {
+					auto_arcs
+				}
+			}
+			MaximizeArcs::On => self.arcs_helper(0., 1., error, max_iterations, false).0,
+			MaximizeArcs::Off => self
+				.get_extrema_t_list()
+				.windows(2)
+				.flat_map(|t_pair| self.arcs_helper(t_pair[0], t_pair[1], error, max_iterations, false).0)
+				.collect::<Vec<CircleArc>>(),
+		}
+	}
+
+	/// Implements an algorithm that approximates a bezier curve with circular arcs.
+	/// This algorithm uses a method akin to binary search to find an arc that approximates a maximal segment of the curve.
+	/// Once a maximal arc has been found for a sub-segment of the curve, the algorithm continues by starting again at the end of the previous approximation.
+	/// More details can be found in the [Approximating a Bezier curve with circular arcs](https://pomax.github.io/bezierinfo/#arcapproximation) section of Pomax's bezier curve primer.
+	/// A caveat with this algorithm is that it is possible to find erroneous approximations in cases such as in a very narrow `U`.
+	/// - `stop_when_invalid`: Used to determine whether the algorithm should end early erroneous approximations are encountered.
+	///
+	/// Returns a tuple where the first element is the list of circular arcs and the second is the `t` value where the next segment should start from.
+	/// The second value will be `1.` except for when `stop_when_invalid` is true and an invalid approximation is encountered.
+	fn arcs_helper(&self, local_low: f64, local_high: f64, error: f64, max_iterations: i32, stop_when_invalid: bool) -> (Vec<CircleArc>, f64) {
+		let mut low = local_low;
+		let mut middle = local_low + (local_high - local_low) / 2.;
+		let mut high = local_high;
+		let mut previous_high = local_high;
+
+		let mut iterations = 0;
+		let mut previous_arc = CircleArc::default();
+		let mut was_previous_good = false;
+		let mut arcs = Vec::new();
+
+		// Outer loop to iterate over the curve
+		while low < local_high {
+			if iterations > max_iterations {
+				break;
+			}
+			// Inner loop to find the next maximal segment of the curve that can be approximated with a circular arc
+			loop {
+				if iterations > max_iterations {
+					break;
+				}
+				iterations += 1;
+				let p1 = self.evaluate(low);
+				let p2 = self.evaluate(middle);
+				let p3 = self.evaluate(high);
+
+				// If the segment is linear, move on to next segment
+				if utils::are_points_collinear(p1, p2, p3) {
+					previous_high = high;
+					low = high;
+					high = 1.;
+					middle = low + (high - low) / 2.;
```

Comment:
`middle = (low + high) / 2;`

Reply:
The reason we use `low + (high-low) / 2` instead of (low+high) / 2` is to avoid overlow in the case that `low+high` is larger than the size of a `f64` type.

Reply:
Hadn't thought of that, thanks for pointing it out. Although if `high` is a large number and `low` is a small (high negative) number, that minus sign flips the `low` to be positive and then it's adding two high numbers again and causing overflow within the parentheses. However, it's probably more likely that `low` and `high` will be similar, not wildly far apart, so your solution here is likely more robst. Is it worth the extra few clock cycles? Not sure. But I'd say keep it as is.

Reply:
There's some places in `intersections_between_subcurves` where we use something akin to `(low + high)/2` rather than `low + (high-low)/2`. Should we change those or leave them be?

Reply:
@hannahli2010 Yeah, I had given the same comment to you about changing those to the form `(low + high)/2` since I didn't think about overflow/underflow. I'd say change them back to `low + (high - low) / 2` wherever there's the possibility that user input could give really high or low numbers, but for efficiency keep it with `(low + high) / 2` form in places where it's unlikely or impossible that user input would cause overflow/underflow. Robustness, then efficiency, is our priority. Just remember that both styles will overflow or underflow, it's just less likely with the `low + (high - low) / 2` form because they both have to be large numbers of opposite signs. So please just use your judgement about what seems possible or likely about the user input and be robust where necessary then efficient as a second priority.

Reply:
I think it should be fine to change because for `low` and `high`, we're iterating between the range of 0 and 1 (i.e. we're iterating over the `t` range of a bezier curve.

Reply:
Ah, I was confused about that. Then yes, definitely use `(low + high) / 2` when the range is bounded.